src/dune/command.ml

```diff
@@ -82,30 +72,43 @@ let expand_static_exn ~dir t =
   | Ok res -> res
 
 let expand ~dir ts =
-  let rec loop = function
-    | A s -> Build.return [ s ]
-    | As l -> Build.return l
+  let rec loop : _ -> _ Build.With_targets.t = function
+    | A s -> Build.With_targets.return [ s ]
+    | As l -> Build.With_targets.return l
     | Dep fn ->
-      Build.map (Build.path fn) ~f:(fun () -> [ Path.reach fn ~from:dir ])
-    | Path fn -> Build.return [ Path.reach fn ~from:dir ]
+      Build.no_targets
+        (Build.map (Build.path fn) ~f:(fun () -> [ Path.reach fn ~from:dir ]))
+    | Path fn -> Build.With_targets.return [ Path.reach fn ~from:dir ]
     | Deps fns ->
-      Build.map (Build.paths fns) ~f:(fun () ->
-          List.map fns ~f:(Path.reach ~from:dir))
-    | Paths fns -> Build.return (List.map fns ~f:(Path.reach ~from:dir))
-    | S ts -> Build.map (Build.all (List.map ts ~f:loop)) ~f:List.concat
+      Build.no_targets
+        (Build.map (Build.paths fns) ~f:(fun () ->
+             List.map fns ~f:(Path.reach ~from:dir)))
+    | Paths fns ->
+      Build.With_targets.return (List.map fns ~f:(Path.reach ~from:dir))
+    | S ts ->
+      Build.With_targets.map
+        (Build.With_targets.all (List.map ts ~f:loop))
+        ~f:List.concat
     | Concat (sep, ts) ->
-      Build.map (loop (S ts)) ~f:(fun x -> [ String.concat ~sep x ])
-    | Target fn -> Build.return [ Path.reach (Path.build fn) ~from:dir ]
-    | Dyn dyn -> Build.dyn_deps (Build.map dyn ~f:(expand_static_exn ~dir))
-    | Fail f -> Build.fail f
-    | Hidden_deps deps -> Build.map (Build.deps deps) ~f:(fun () -> [])
-    | Hidden_targets _ -> Build.return []
-    | Expand f -> (
-      match f ~dir with
-      | Error e -> Build.fail e
-      | Ok (args, deps) ->
-        let open Build.O in
-        Build.deps deps >>> Build.return args )
+      Build.With_targets.map (loop (S ts)) ~f:(fun x ->
+          [ String.concat ~sep x ])
+    | Target fn ->
+      Build.add ~targets:[ fn ]
+        (Build.return [ Path.reach (Path.build fn) ~from:dir ])
+    | Dyn dyn ->
+      Build.no_targets
+        (Build.dyn_deps (Build.map dyn ~f:(expand_static_exn ~dir)))
+    | Fail f -> Build.no_targets (Build.fail f)
+    | Hidden_deps deps ->
+      Build.no_targets (Build.map (Build.deps deps) ~f:(fun () -> []))
+    | Hidden_targets fns -> Build.add ~targets:fns (Build.return [])
+    | Expand f ->
```

Comment:
With the type of `expand` changing, the names `Expand` are a bit out of sync now. I don't have any good ideas though. `Expand_no_targets` could work, but we might as well keep `Expand` if you like that more.

Reply:
I've updated the comment in `command.mli` to make it clearer that `expand` does not include target declarations.

Reply:
Looks good.